From c02aa121a03e0bf27f7e3144147135aedbc0be8c Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Fri, 26 Apr 2024 15:04:49 +0800
Subject: [PATCH 19/67] PHYTIUM: net/phytmac: Roll back rx_clean version

Roll back rx_clean version to not use batching.

Mainline: NA
Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: Ie6e7585304ac6a7b9101e04c5af5219b3a361996

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/1040f4e7b52143571a6fa9e85e13cd0c1642aafc
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/net/ethernet/phytium/phytmac.h      |  2 --
 drivers/net/ethernet/phytium/phytmac_main.c | 24 ++++++++-------------
 drivers/net/ethernet/phytium/phytmac_v1.c   | 23 ++++++--------------
 3 files changed, 16 insertions(+), 33 deletions(-)

diff --git a/drivers/net/ethernet/phytium/phytmac.h b/drivers/net/ethernet/phytium/phytmac.h
index b90e1551499e..33c0fc78415b 100644
--- a/drivers/net/ethernet/phytium/phytmac.h
+++ b/drivers/net/ethernet/phytium/phytmac.h
@@ -38,7 +38,6 @@
 #define MIN_TX_RING_SIZE		64
 #define MIN_RX_RING_SIZE		64
 #define DEFAULT_TX_DESC_MIN_FREE	64
-#define DEFAULT_RX_DESC_MIN_FREE	64

 #define MEMORY_SIZE			4096
 #define MHU_SIZE			0x20
@@ -499,7 +498,6 @@ struct phytmac_hw_if {
 			       struct packet_info *packet);
 	void (*init_rx_map)(struct phytmac_queue *queue, u32 index);
 	unsigned int (*rx_map)(struct phytmac_queue *queue, u32 index, dma_addr_t addr);
-	unsigned int (*rx_clean)(struct phytmac_queue *queue, u32 cleaned_count);
 	void (*transmit)(struct phytmac_queue *queue);
 	void (*restart)(struct phytmac *pdata);
 	int (*tx_complete)(const struct phytmac_dma_desc *desc);
diff --git a/drivers/net/ethernet/phytium/phytmac_main.c b/drivers/net/ethernet/phytium/phytmac_main.c
index 28b6dc86f668..0053412c2632 100644
--- a/drivers/net/ethernet/phytium/phytmac_main.c
+++ b/drivers/net/ethernet/phytium/phytmac_main.c
@@ -793,16 +793,12 @@ static void phytmac_rx_clean(struct phytmac_queue *queue)
 	unsigned int index, space;
 	dma_addr_t paddr;
 	struct sk_buff *skb;
-	unsigned int rx_unclean = 0;

 	space = CIRC_SPACE(queue->rx_head, queue->rx_tail,
 			   pdata->rx_ring_size);

-	if (space < DEFAULT_RX_DESC_MIN_FREE)
-		return;
-
-	index = queue->rx_head & (pdata->rx_ring_size - 1);
 	while (space > 0) {
+		index = queue->rx_head & (pdata->rx_ring_size - 1);
 		if (!queue->rx_skb[index]) {
 			skb = netdev_alloc_skb(pdata->ndev, pdata->rx_buffer_len);
 			if (unlikely(!skb)) {
@@ -811,7 +807,7 @@ static void phytmac_rx_clean(struct phytmac_queue *queue)
 			}

 			paddr = dma_map_single(pdata->dev, skb->data,
-						pdata->rx_buffer_len, DMA_FROM_DEVICE);
+					       pdata->rx_buffer_len, DMA_FROM_DEVICE);
 			if (dma_mapping_error(pdata->dev, paddr)) {
 				dev_kfree_skb(skb);
 				break;
@@ -820,21 +816,19 @@ static void phytmac_rx_clean(struct phytmac_queue *queue)
 			queue->rx_skb[index] = skb;

 			hw_if->rx_map(queue, index, paddr);
+		} else {
+			hw_if->rx_map(queue, index, 0);
 		}

-		index = (index + 1) & (pdata->rx_ring_size - 1);
-		rx_unclean++;
+		queue->rx_head++;
+		if (queue->rx_head >= pdata->rx_ring_size)
+			queue->rx_head &= (pdata->rx_ring_size - 1);
+
 		space--;
 	}

 	/* make newly descriptor to hardware */
 	wmb();
-	hw_if->rx_clean(queue, rx_unclean);
-	/* make newly descriptor to hardware */
-	wmb();
-	queue->rx_head += rx_unclean;
-	if (queue->rx_head >= pdata->rx_ring_size)
-		queue->rx_head &= (pdata->rx_ring_size - 1);
 }

 static int phytmac_rx(struct phytmac_queue *queue, struct napi_struct *napi,
@@ -1203,7 +1197,7 @@ static unsigned int phytmac_tx_map(struct phytmac *pdata,
 {
 	dma_addr_t mapping;
 	struct phytmac_hw_if *hw_if = pdata->hw_if;
-	unsigned int len, i, tx_tail = queue->tx_tail;
+	unsigned int len, i, tx_tail;
 	struct phytmac_tx_skb *tx_skb = NULL;
 	unsigned int offset, size, count = 0;
 	unsigned int f, nr_frags = skb_shinfo(skb)->nr_frags;
diff --git a/drivers/net/ethernet/phytium/phytmac_v1.c b/drivers/net/ethernet/phytium/phytmac_v1.c
index b96547e54d82..b823a07a731b 100644
--- a/drivers/net/ethernet/phytium/phytmac_v1.c
+++ b/drivers/net/ethernet/phytium/phytmac_v1.c
@@ -927,22 +927,14 @@ static unsigned int phytmac_rx_map_desc(struct phytmac_queue *queue,
 			addr |= PHYTMAC_BIT(RX_WRAP);
 		desc->desc1 = 0;
 		desc->desc2 = upper_32_bits(addr);
-		desc->desc0 = lower_32_bits(addr) | PHYTMAC_BIT(RX_USED);
-	}
-	return 0;
-}
-
-static unsigned int phytmac_rx_clean_desc(struct phytmac_queue *queue, u32 count)
-{
-	struct phytmac_dma_desc *desc;
-	u32 index = queue->rx_head + count - 1;
-
-	while (count) {
-		desc = phytmac_get_rx_desc(queue, index);
-		desc->desc0 &= ~PHYTMAC_BIT(RX_USED);
+		/* Make newly descriptor to hardware */
+		dma_wmb();
+		desc->desc0 = lower_32_bits(addr);
+	} else {
+		desc->desc1 = 0;
+		/* Make newly descriptor to hardware */
 		dma_wmb();
-		index--;
-		count--;
+		desc->desc0 &= ~PHYTMAC_BIT(RX_USED);
 	}

 	return 0;
@@ -1386,7 +1378,6 @@ struct phytmac_hw_if phytmac_1p0_hw = {
 	.get_desc_addr = phytmac_get_desc_addr,
 	.init_rx_map = phytmac_init_rx_map_desc,
 	.rx_map = phytmac_rx_map_desc,
-	.rx_clean = phytmac_rx_clean_desc,
 	.rx_checksum = phytmac_rx_checksum,
 	.rx_single_buffer = phytmac_rx_single_buffer,
 	.rx_pkt_start = phytmac_rx_sof,
--
2.50.1
